### PR TITLE
[GUILD-3229] - Farcaster as identity

### PR DESCRIPTION
**Pull request by cs-balazs**
[Linear](https://linear.app/guildxyz/issue/GUILD-3229/fe-implementation)

- Adds a Farcaster connection & disconnection flow
- On connection, the user has presented a QR code, which, after scanning, redirects to Warpcast, where the connection request can be approved
  - The FE polls the state of this approval from the BE
- Once the request is approved, the appropriate SWR state is mutated to contain the Farcaster profile
- Disconnect happens with a single DELETE call & appropriate SWR mutation
- Some changes on this branch are fixes for the new tsc settings: https://github.com/guildxyz/guild.xyz/pull/1301/commits/79d28fcf3b4458cc5e7751eb2e84debf4dc8baf4

### CODE REVIEW

File: src/platforms/components/FarcasterIcon.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { Icon } from "@chakra-ui/react"
+import { IconProps } from "phosphor-react"
+import { ForwardedRef, forwardRef } from "react"
+
+function FarcasterIcon(_: IconProps, ref: ForwardedRef<SVGSVGElement>) {
```
Comment (BrickheadJohnny) [inline src/platforms/components/FarcasterIcon.tsx:5]: I wonder if we could just put this icon inside the `static` folder and import it just like we import e.g. the user avatar icons?

File: src/platforms/Farcaster/index.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import FarcasterIcon from "platforms/components/FarcasterIcon"
+import { PlatformAsRewardRestrictions, RewardData } from "platforms/types"
+
+export default {
+  icon: FarcasterIcon,
+  imageUrl: "/platforms/poap.png",
```
Comment (BrickheadJohnny) [inline src/platforms/Farcaster/index.ts:6]: whoopsie, wen poap?

File: src/hooks/useCountdownSeconds.ts [hunk 3]
```diff
@@ -1,27 +1,44 @@
 import { useState } from "react"
 
-const useCountdownSeconds = (seconds: number) => {
-  const [sec, setSec] = useState(seconds)
+const useCountdownSeconds = (seconds?: number, onCountdown?: () => void) => {
```
Comment (BrickheadJohnny) [inline src/hooks/useCountdownSeconds.ts:3]: This hook starts to get more and more complex hehe, what do you think, should we just use the [`useCountdown`](https://usehooks-ts.com/react-hook/use-countdown) hook from `usehooks-ts`? I think we'll add this package to our app anyways, since we'll need some hooks of it when we migrate from Chakra. That way we wouldn't really need to maintain these kind of hooks, just use the premade ones from this package.

File: src/components/create-guild/MultiPlatformGrid/components/MultiPlatformSelectButton.tsx
Comment (BrickheadJohnny) [inline src/components/create-guild/MultiPlatformGrid/components/MultiPlatformSelectButton.tsx:1]: Oh, this component will be removed in https://github.com/guildxyz/guild.xyz/pull/1309, so we don't really need to update it. 😉  (The linked PR will be merged later today)

File: .env
Comment (BrickheadJohnny) [inline .env:1]: Don't forget to revert the changes before merging the PR 😉

File: src/components/common/Layout/components/Account/components/AccountModal/components/SocialAccount/FarcasterProfile.tsx [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+import DisconnectAccountButton from "./components/DisconnectAccountButton"
+import SocialAccountUI from "./components/SocialAccountUI"
+
+// Polling will only start after this much time
+const APPROVAL_POLL_INITIAL_DELAY_MS = 15_000
+// Delay between polls
+const APPROVAL_POLL_INTERVAL_MS = 5000
+// Poll ends, and a new QR is generated this much before the actual deadline
+const POLL_EARLY_END_SEC = 120
+// Enable the manual QR regenation button, when less than this muc seconds is left
```
Comment (BrickheadJohnny) [inline src/components/common/Layout/components/Account/components/AccountModal/components/SocialAccount/FarcasterProfile.tsx:47]: nit: typo

File: src/components/common/Layout/components/Account/components/AccountModal/components/SocialAccount/FarcasterProfile.tsx [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+      <Modal isOpen={isOpen} onClose={handleOnClose} size="lg">
+        <ModalOverlay />
+        <ModalContent>
+          <ModalCloseButton />
+          <ModalHeader pb={0} pr={16} display={"flex"} gap={1} alignItems={"center"}>
+            {isMobile ? (
+              "Connect Farcaster"
+            ) : (
+              <>
+                <Icon boxSize={6} as={DeviceMobileCamera} /> Scan With your phone
```
Comment (BrickheadJohnny) [inline src/components/common/Layout/components/Account/components/AccountModal/components/SocialAccount/FarcasterProfile.tsx:182]: We might need to wrap the text in a span, not sure if translator plugins can handle the title this way

**Review by BrickheadJohnny — COMMENTED** [review]
Left a couple of comments, but fantastic PR overall! 👑

File: src/platforms/Farcaster/index.ts (hunk 2 above)
Comment (cs-balazs) [reply src/platforms/Farcaster/index.ts]: Nice catch! Changed in https://github.com/guildxyz/guild.xyz/pull/1301/commits/d8fc1742ada209187fbfe9365ed29f4ca5815790 👍

File: src/platforms/components/FarcasterIcon.tsx (hunk 1 above)
Comment (cs-balazs) [reply src/platforms/components/FarcasterIcon.tsx]: Great idea! Change: https://github.com/guildxyz/guild.xyz/pull/1301/commits/a12328008b154470df1916fe2b61aabfc0d0719c

File: .env
Comment (cs-balazs) [reply .env:1]: Jeje, reverted to prod now: https://github.com/guildxyz/guild.xyz/pull/1301/commits/5ee1ea200b6cfaaa6d49d0286103e3a920ae7e49

File: src/components/common/Layout/components/Account/components/AccountModal/components/SocialAccount/FarcasterProfile.tsx (hunk 4 above)
Comment (cs-balazs) [reply src/components/common/Layout/components/Account/components/AccountModal/components/SocialAccount/FarcasterProfile.tsx]: 2 typos on one line, nice 😄 
Fixed in https://github.com/guildxyz/guild.xyz/pull/1301/commits/7acc0364a729a924ce682a94bc3ef472f18548a4 but ended up removing this const in https://github.com/guildxyz/guild.xyz/pull/1301/commits/0321fb71d7c04e39b672601fd2599b9c424b1c7c

File: src/components/common/Layout/components/Account/components/AccountModal/components/SocialAccount/FarcasterProfile.tsx (hunk 5 above)
Comment (cs-balazs) [reply src/components/common/Layout/components/Account/components/AccountModal/components/SocialAccount/FarcasterProfile.tsx]: Thanks for pointing that out! Wrapped in a Text in https://github.com/guildxyz/guild.xyz/pull/1301/commits/0f766935d25a9b2969fea9d53b82d569a256b691

File: src/components/create-guild/MultiPlatformGrid/components/MultiPlatformSelectButton.tsx
Comment (cs-balazs) [reply src/components/create-guild/MultiPlatformGrid/components/MultiPlatformSelectButton.tsx:1]: Nice, thanks! Reverted most of the changes in https://github.com/guildxyz/guild.xyz/pull/1301/commits/79249a0b35b1fa63306265593589755d71c2538e

File: src/hooks/useCountdownSeconds.ts (hunk 3 above)
Comment (cs-balazs) [reply src/hooks/useCountdownSeconds.ts]: Nice, added the lib & changed the hook usage in https://github.com/guildxyz/guild.xyz/pull/1301/commits/0321fb71d7c04e39b672601fd2599b9c424b1c7c
I didn't remove the hook yet, but I'll do that in a follow-up PR :ayy:

One thing, which I'm not sure is that the deadline is a little less accurate since we need to pass it in the hook call, but it shouldn't be a problem, just something to keep in mind

**Review by BrickheadJohnny — APPROVED** [review]
